F/Faiss/common.jl

```diff
@@ -0,0 +1,102 @@
+# Collection of sources required to complete build
+sources = [
+    GitSource("https://github.com/facebookresearch/faiss.git", "d243e628880676332263347817b3fe7f474b8b5b"),
+    DirectorySource(joinpath(@__DIR__, "bundled")),
+]
+
+# Bash recipe for building across all platforms
+script = raw"""
+# Needs CMake >= 3.23.1 provided via HostBuildDependency
+apk del cmake
+
+cd faiss
+
+atomic_patch -p1 ../patches/faiss-mingw32-cmake.patch
+atomic_patch -p1 ../patches/faiss-mingw32-InvertedListsIOHook.patch
+atomic_patch -p1 ../patches/faiss-mingw32.patch
+
+cmake_extra_args=()
+
+if [[ $bb_full_target == *cuda* ]]; then
+    cuda_version=${bb_full_target##*-cuda+}
+    if [[ $cuda_version == "11.8" ]]; then
+        cuda_archs="60-real;61-real;62-real;70-real;72-real;75-real;80;86-real;87;89;90"
+    elif [[ $cuda_version == "12.1" ]]; then
+        cuda_archs="70-real;72-real;75-real;80;86-real;87;89;90"
```

Comment:
@maleadt Which CUDA archs are sensible to build for? (and what does `-real` mean?)
Source: https://github.com/facebookresearch/faiss/blob/v1.9.0/.github/workflows/build.yml#L182-L260

Shouldn't PTX be added?

Would it make sense to add a `default_archs(cuda_version)` method to the CUDA module?

I tried CUDA 12.6 for SM100 support, but it failed with: `nvcc fatal   : Unsupported gpu architecture 'compute_100'` (?).

Reference: https://arnon.dk/matching-sm-architectures-arch-and-gencode-for-various-nvidia-cards/

Reply:
`ptxas` from CUDA 12.6 supports up to `sm_90a`.

I don't think we can provide a reasonable default here, as it depends on the kinds of kernels being used in the application.

`-real` comes from CMake, https://cmake.org/cmake/help/latest/prop_tgt/CUDA_ARCHITECTURES.html#prop_tgt:CUDA_ARCHITECTURES, but the meaning comes from ptxas/nvcc: real architectures generate object code, virtual architectures generate PTX code.

Reply:
OK.

So, IIUC, PTX provides some form of code that can be compiled to the given device architecture at run-time, right? For a certain CUDA compute capability major version, e.g. for 8.0, specializing for sm_89 based on PTX code for sm_80.

Would it thus make sense for faiss, where sm_90 seems relevant to include, but not so much sm_90a (sounds like its only needed for some specific features) to specify (CMake) `"CUDA_ARCHITECTURES=...;90;100-virtual"`? (which would then include 90-virtual for possible execution on sm_90a ... and PTX code for sm_100?)

Reply:
Wrt. "default_archs": I was thinking that perhaps you could avoid answering this question more than once - or at least fewer times :-) Maybe with `CUDA.unreasonable_default_archs(cuda_version; include_jetson = true)` :-) Given that some might think (a few years ago), "I want to have Ada Lovelace-support, so I'm building for CUDA 11.8", and then have an (unreasonable) default starting point that includes 8.9 (real+virtual), and some earlier versions.

Reply:
> PTX provides some form of code that can be compiled to the given device architecture at run-time

Correct.



> PTX code for sm_100

`ptxas` from CUDA 12.6 doesn't support compute capability 10.



> Wrt. "default_archs": I was thinking that perhaps you could avoid answering this question more than once - or at least fewer times

It's not unreasonable, but just a bit of work to generalize across build systems. You're using CMake here, where the architecture flags are different from what's used by `nvcc` (which by simpler build systems is used directly).